File: src/main/java/com/binance/api/client/domain/event/CandlestickEvent.java
package com.binance.api.client.domain.event;

import com.binance.api.client.domain.market.Candlestick;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

/**
 * An interval candlestick for a symbol providing informations on price that can be used to produce candlestick charts.
 */
@JsonDeserialize(using = CandlestickEventDeserializer.class)
public class CandlestickEvent extends Candlestick {
  public CandlestickEvent(String eventType,
                          long eventTime,
                          String symbol,
                          Long openTime,
                          String open,
                          String high,
                          String low,
                          String close,
                          String volume,
                          Long closeTime,
                          String intervalId,
                          Long firstTradeId,
                          Long lastTradeId,
                          String quoteAssetVolume,
                          Long numberOfTrades,
                          String takerBuyBaseAssetVolume,
                          String takerBuyQuoteAssetVolume,
                          Boolean isBarFinal) {
    super(openTime, open, high, low, close, volume, closeTime, quoteAssetVolume, numberOfTrades, takerBuyBaseAssetVolume, takerBuyQuoteAssetVolume);
    this.eventType = eventType;
    this.eventTime = eventTime;
    this.symbol = symbol;
    this.intervalId = intervalId;
    this.firstTradeId = firstTradeId;
    this.lastTradeId = lastTradeId;
    this.isBarFinal = isBarFinal;
  }

  public final String eventType;
  public final long eventTime;
  public final String symbol;
  public final String intervalId;
  public final Long firstTradeId;
  public final Long lastTradeId;
  public final Boolean isBarFinal;
}

File: src/main/java/com/binance/api/client/domain/event/CandlestickEvent.scala
package com.binance.api.client.domain.event

import com.binance.api.client.domain._

/**
  * An interval candlestick for a symbol providing information on price that can be used to produce candlestick charts.
  */
case class CandlestickEvent(
    eventType:   String,
    eventTime:   Instant,
    symbol:      Symbol,
    candlestick: CandlestickDetailed,
)

case class CandlestickDetailed(
    openTime:                 Instant,
    open:                     Price,
    high:                     Price,
    low:                      Price,
    close:                    Price,
    volume:                   Volume,
    closeTime:                Instant,
    intervalId:               String,
    firstTradeId:             Long,
    lastTradeId:              Long,
    quoteAssetVolume:         String,
    numberOfTrades:           Long,
    takerBuyBaseAssetVolume:  String,
    takerBuyQuoteAssetVolume: String,
    isBarFinal:               Boolean
) extends CandlestickBase
